internal_transfer_with_agreed_amount/wizards/internal_transfer_multicurrency.py

```diff
@@ -0,0 +1,76 @@
+        if not active_ids or active_model != "account.payment":
+            return False
+        payment = self.env["account.payment"].browse(active_ids)
+        return payment.destination_journal_id.currency_id or payment.company_currency_id
+
+    def _validate_currencies_in_internal_transfer(self, payment):
+        payment_currency = payment.currency_id
+        company_currency = payment.company_currency_id
+        send_currency = payment.journal_id.currency_id or company_currency
+        receive_currency = self.currency_id
+        currencies = payment_currency | company_currency | send_currency | receive_currency
+        if len(currencies) == 2:
+            return currencies - payment_currency
+        raise ValidationError(
+            _(
+                """
+                Agreed amount is available when there are two considered currencies.
+                Currently, the %s currencies are considered for this process.
+                """,
+                ", ".join(currencies.mapped("name")),
+            )
+        )
+
+    def apply(self):
+        active_ids = self._context.get("active_ids")
+        active_model = self._context.get("active_model")
+        if not active_ids or active_model != "account.payment":
+            return
+        payment = self.env["account.payment"].browse(active_ids)
+        exchange_currency = self._validate_currencies_in_internal_transfer(payment)
+        payment.filtered(lambda p: p.state == "draft").action_post()
+        amls = (payment | payment.paired_internal_transfer_payment_id).line_ids
+        if not self.agreed_amount:
+            return
+        to_reconcile = amls.filtered(
+            lambda l: l.account_id == l.company_id.transfer_account_id
+        ).full_reconcile_id.reconciled_line_ids
+        if exchange_currency == payment.company_currency_id:
+            if len(amls) != 4:
+                return
+            amls.move_id.button_draft()
+            debit = amls.filtered(lambda x: x.debit > 0)
+            credit = amls.filtered(lambda x: x.credit > 0)
+            debit.with_context(check_move_validity=False).write({"debit": self.agreed_amount})
+            credit.with_context(check_move_validity=False).write({"credit": self.agreed_amount})
+            amls.move_id.action_post()
+            to_reconcile.reconcile()
+            return
+        aml_positive = amls.filtered(lambda a: a.amount_currency > 0.00)
+        aml_negative = amls.filtered(lambda a: a.amount_currency < 0.00)
+        if len(aml_positive) > 2 or len(aml_negative) > 2:
+            return
+        aml_positive.move_id.button_draft()
+        context = {"check_move_validity": False, "transfer_with_agreed_amount": True}
+        aml_positive.with_context(**context).write({"amount_currency": self.agreed_amount})
+        aml_negative.with_context(**context).write({"amount_currency": -self.agreed_amount})
+        aml_positive.move_id.with_context(transfer_with_agreed_amount=True).action_post()
```

Comment:
Have you tried to use aml_record._compute_amount_residual()

in order to avoid the inheritance upwards?

Reply:
So far, let us go this Way.

Reply:
What I had done here was a mess, but I redeemed my sins